File: leetcode/gas_station/GasStation.cpp
class Solution {
public:
    int canCompleteCircuit(vector<int> &gas, vector<int> &cost) {
        int n = gas.size(); 
        int start = 0; 
        int sum = 0; 
        int total = 0; 
        
        for (int i = 0; i < n; i++) {
            int delta = gas[i] - cost[i]; 
            total += delta; 
            sum += delta; 
            
            // Current segment consumes gas
            // (Won't make from start to i)
            if (sum < 0) {  
                sum = 0;  // Start a new segment
                start = i + 1; // Start a new starting station for this new segment
            }
        }
        
        return total >= 0 ? start : -1; 
    }
};

File: leetcode6/gas_station.java
/**
 * Find the starting point where the car can go around the circle. 
 * The idea was to find the last end point of a segment where the total gas 
 * of that segment is less than the total cost of that segment. 
 */ 
public class Solution {
    public int canCompleteCircuit(int[] gas, int[] cost) {
        int totalGas = 0; 
        int totalCost = 0; 
        int len = gas.length; 
        
        for (int i = 0; i < len; i++) {
            totalGas += gas[i]; 
            totalCost += cost[i]; 
        }
        
        if (totalGas < totalCost) {
            return -1; 
        }
        
        int currGas = 0; 
        int currCost = 0; 
        int result = 0; 
        
        for (int i = 0; i < len; i++) {
            if ((currGas + gas[i]) < (currCost + cost[i])) {
                result = i + 1; 
                currGas = 0; 
                currCost = 0; 
            } else {
                currGas += gas[i]; 
                currCost += cost[i]; 
            }
        }
        
        return result;
    }
}
